src/thread_event.c

```diff
@@ -0,0 +1,117 @@
+#define JEMALLOC_THREAD_EVENT_C_
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/thread_event.h"
+
+static void
+thread_event_adjust_thresholds_helper(tsd_t *tsd, uint64_t last_event,
+    uint64_t wait) {
+	assert(last_event == tsd_thread_allocated_last_event_get(tsd));
+	assert(wait > 0);
+	uint64_t threshold = last_event + wait;
+	*tsd_thread_allocated_thresholdp_get(tsd) = threshold;
+	if (likely(last_event < threshold &&
+	    threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX)) {
+		*tsd_thread_allocated_threshold_fastp_get(tsd) = threshold;
+	} else {
+		/*
+		 * threshold either overflowed or exceeded
+		 * THREAD_ALLOCATED_THRESHOLD_FAST_MAX.  Zero
+		 * thread_allocated_threshold_fast so that
+		 * thread_event_fast() will always trigger.
+		 */
+		*tsd_thread_allocated_threshold_fastp_get(tsd) = 0;
+	}
+}
+
+void
+thread_event_trigger(tsd_t *tsd, uint64_t usize, bool next_time) {
+	/* usize has already been added to thread_allocated. */
+	uint64_t thread_allocated_after = tsd_thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t bulk_transfer = thread_allocated_after -
+	    tsd_thread_allocated_last_event_get(tsd);
+
+	if (bulk_transfer < usize) {
+		/* bulk_transfer overflowed. */
+		bulk_transfer = UINT64_MAX;
+	}
+
+	/*
+	 * In case none of the events has its condition on, the UINT64_MAX
+	 * initialization will automatically delay the next event as much as
+	 * possible.
+	 */
+	uint64_t wait = UINT64_MAX;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait =					\
+		    tsd_##event##_event_wait_get(tsd);			\
+		assert(event_wait > 0);					\
+		if (event_wait <= bulk_transfer) {			\
+			event_wait = 0;					\
+		} else {						\
+			event_wait -= bulk_transfer;			\
+		}							\
+		if (unlikely(next_time || !tsd_nominal(tsd) ||		\
+		    tsd_reentrancy_level_get(tsd) > 0) &&		\
+		    event_wait == 0) {					\
+			/* Delay the event. */				\
+			event_wait = 1;					\
+		}							\
+		*tsd_##event##_event_waitp_get(tsd) = event_wait;	\
```

Comment:
The logic above probably deserves to be separated into a non macro static function, using the same trick we mentioned before. Also next_time rename to delay_event.

Reply:
Well it's doable but it might end up making the code more sophisticated than necessary. E.g. the function must accept quite a few parameters: `tsd`, pointer to the individual wait time, pointer to the `wait` variable, the `bulk_transfer` variable, and the `delay_event` flag. Also it seems tricky to name the function - it's doing several things: decrease (and possibly delay) wait time, and get the global minimal wait time.

If we really don't like these long stretches of macro-defined logic, how about this: I can pull the logic of getting the minimal wait time into a separate function (which is also used multiple times else where), and pull the logic of setting the new wait time I'm now having in this `thread_event_trigger()` function (not including the part for obtaining the minimal wait time) into another separate function. Then in this `thread_event_trigger()` function I'll call these two utility functions separately. The only downside is that the individual wait times need to be stored and read again, but we're on the slow path anyway. I think in this way the utility functions can become cleaner and more modular. Thoughts?

Reply:
I see your points. Let’s keep it in macro for now.